bundle/edu.gemini.lchquery.servlet/src/main/scala/edu/gemini/lchquery/servlet/LchQueryFunctor.scala

```diff
@@ -0,0 +1,230 @@
+package edu.gemini.lchquery.servlet
+
+import java.io.Serializable
+import java.security.Principal
+import java.text.SimpleDateFormat
+import java.util
+import java.util.{Date, TimeZone}
+import java.util.logging.{Level, Logger}
+
+import edu.gemini.odb.browser._
+import edu.gemini.pot.sp.{ISPNode, ISPObservation, ISPProgram}
+import edu.gemini.pot.spdb.{DBAbstractQueryFunctor, IDBDatabaseService, IDBFunctor, IDBParallelFunctor}
+import edu.gemini.skycalc.{DDMMSS, HHMMSS}
+import edu.gemini.spModel.gemini.obscomp.SPSiteQuality
+import edu.gemini.spModel.obs.{ObsClassService, ObservationStatus}
+import edu.gemini.spModel.rich.shared.immutable._
+import edu.gemini.spModel.target.SPTarget
+import edu.gemini.spModel.target.env.TargetEnvironment
+import edu.gemini.spModel.too.Too
+
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Success, Try}
+
+
+class LchQueryFunctor(queryType: LchQueryFunctor.QueryType,
+                      programParams: List[(LchQueryParam[ISPProgram], String)],
+                      observationParams: List[(LchQueryParam[ISPObservation], String)])
+  extends DBAbstractQueryFunctor with IDBParallelFunctor {
+
+  // The current QueryResult. Must be mutable.
+  var queryResult: QueryResult = new QueryResult() {
+    setProgramsNode(new ProgramsNode())
+  }
+
+  // Called once per program by IDBQueryRunner implementation.
+  override def execute(db: IDBDatabaseService, node: ISPNode, principals: util.Set[Principal]): Unit = {
+    val prog = node.asInstanceOf[ISPProgram]
+
+    def programMatches: Boolean = programParams.forall {
+      case (LchQueryParam(_, valueMatcher), paramValue) => valueMatcher.matches(paramValue, prog)
+    }
+    def observationMatches(obs: ISPObservation): Boolean = observationParams.forall {
+      case (LchQueryParam(_, valueMatcher), paramValue) => valueMatcher.matches(paramValue, obs)
+    }
+
+    Try {
+      // See if an ISPProgram matches the query specifications.
+      if (programMatches) {
+        val matchingObs = prog.getAllObservations.asScala.toList.filter(observationMatches)
+        if (matchingObs.nonEmpty) {
```

Comment:
Yes. At least, that is how things worked in the previous code. I'll indicate the relevant section with an additional comment and flag you in it. It seems, though, that if there are no observations / no matching observations, the program is simply skipped.